ds701_book/20-Recommender-Systems.qmd

```diff
@@ -127,174 +128,230 @@ Or it can be done passively:
 
 ```{python}
-#| hide_input: true
-#| slideshow: {slide_type: '-'}
-#| tags: [remove-input]
 reviews_per_movie = df.groupby('ProductId').count()['Id'].values
 frac_below_mean = np.sum(reviews_per_movie < (n_reviews/n_movies))/len(reviews_per_movie)
-plt.plot(sorted(reviews_per_movie), '.-')
+plt.plot(sorted(reviews_per_movie, reverse=True), '.-')
 xmin, xmax, ymin, ymax = plt.axis()
 plt.hlines(n_reviews/n_movies, xmin, xmax, 'r', lw = 3)
 plt.ylabel('Number of Ratings', fontsize = 14)
 plt.xlabel('Movie', fontsize = 14)
+plt.legend(['Number of Ratings', 'Average Number of Ratings'], fontsize = 14)
 plt.title(f'Amazon Movie Reviews\nNumber of Ratings Per Movie\n' +
           f'{frac_below_mean:0.0%} of Movies Below Average', fontsize = 16);
 ```
 
+---
+
 Likewise, although the average user writes 14 reviews, almost all users write even fewer reviews.
 
 ```{python}
-#| hide_input: true
-#| tags: [remnove-input]
 reviews_per_user = df.groupby('UserId').count()['Id'].values
 frac_below_mean = np.sum(reviews_per_user < (n_reviews/n_users))/len(reviews_per_user)
-plt.plot(sorted(reviews_per_user), '.-')
+plt.plot(sorted(reviews_per_user, reverse=True), '.-')
 xmin, xmax, ymin, ymax = plt.axis()
 plt.hlines(n_reviews/n_users, xmin, xmax, 'r', lw = 3)
 plt.ylabel('Number of Ratings', fontsize = 14)
 plt.xlabel('User', fontsize = 14)
+plt.legend(['Number of Ratings', 'Average Number of Ratings'], fontsize = 14)
 plt.title(f'Amazon Movie Reviews\nNumber of Ratings Per User\n' +
           f'{frac_below_mean:0.0%} of Users Below Average', fontsize = 16);
 ```
 
+## Objective Function
+
+Ultimately, our goal is to predict the rating that a user would give to an item.
+
+For that, we need to define a loss or objective function.
+
 A typical objective function is root mean square error (RMSE)
 
-$$ \text{RMSE} = \sqrt{1/|S| \sum_{(i,u)\in S} (\hat{r}_{ui} - r_{ui})^2} $$
+$$ 
+\text{RMSE} = \sqrt{\frac{1}{|S|} \sum_{(i,u)\in S} (\hat{r}_{ui} - r_{ui})^2} 
+$$
+
+where 
+
+* $r_{ui}$ is the rating that 
+* user $u$ gives to 
+* item $i$, and 
+* $S$ is the set of all ratings.
 
-where $ r_{ui} $ is the rating that user $u$ gives to item $i$, and $S$ is the set of all ratings.
+---
 
 OK, now we know the problem and the data available.   How can we address the problem?
 
 The earliest method developed is called __collaborative filtering.__
 
+# Colaborative Filtering
+
 ## Collaborative Filtering
 
-The central idea of collaborative filtering is that the set of known recommendations can be considered to be a __bipartite graph.__
+The central idea of collaborative filtering is that the set of known 
+recommendations can be considered to be a __bipartite graph.__
 
-<center>
-    
-<img src="figs/L20-bipartite.png" alt="Figure" width="35%">
-    
-</center>
+![](figs/L20-bipartite.png){width="35%" fig-align="center"}
 
 The nodes of the bipartite graph are __users__ and __items__.   
 
 Each edge corresponds to a known rating $r_{ui}.$
 
+---
+
```

Comment:
I think we should add another sentence to explain how we get a recommendation by traversing the bipartite graph. It's not completely clear to me how the people in the figure are related to the recommendations.

Reply:
that's what is explained in the following slides, no?

Reply:
Yes. I just found the figure confusing because it seems Joe is only connected to ratings through other customers. I thought with the previous bipartite graph that Joe would be linked to the ratings himself. We can leave it as is. I figure the intent is to convey that Joe can get ratings based on what he views and then the corresponding ratings of others who watched similar movies.

Reply:
indeed, i looked at the original lecture slides from UCB and it goes into more explanation... we're just using it as a kind of eye candy I guess.